### PR TITLE
Architecture: Cleanup Type System + Schema Partition

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Refactored Type Checker and added a Schema Partition

## Details
* Create dedicated partition (`PartitionNum 1u8`) for all `ScryptoSchema` to be stored, other system partitions have been moved down by 1 (e.g. `Metadata` partition moved from `1u8` to `2u8`
* Add system capability to map an application layer `CollectionIndex` to a physical partition number. This is used so that a Package may remain in the application layer
* Add notion of `TypeSubstitutionRef` which is the reference to the type to replace some generic argument
* Use the "event name" rather than a `TypePointer` for `EventTypeIdentifier`
* Move reference/ownership schema checks into sbor validation of payload
* Consolidated all type checking into the system layer `system_type_checker.rs`
* `IndexedStateSchema` in the `BlueprintDefinition` substate has been slightly updated to use an enum `PartitionDescription` rather than just a u8

## Update Recommendations
`IndexedStateSchema` as a part of the `BlueprintDefinition` substate has been updated from:
```rust
pub struct IndexedStateSchema {
    pub fields: Option<(PartitionOffset, Vec<FieldSchema<TypePointer>>)>,
    pub collections: Vec<(PartitionOffset, BlueprintCollectionSchema<TypePointer>)>,
    pub num_partitions: u8,
}
```

to:

```rust
pub struct IndexedStateSchema {
    pub fields: Option<(PartitionDescription, Vec<FieldSchema<BlueprintPayloadDef>>)>,
    pub collections: Vec<(
        PartitionDescription,
        BlueprintCollectionSchema<BlueprintPayloadDef>,
    )>,
    pub num_logical_partitions: u8,
}
```

### CODE REVIEW

File: radix-engine-interface/src/types/event_id.rs [hunk 1]
```diff
@@ -16,7 +16,7 @@ use utils::ContextualDisplay;
 /// It is important to note that application events are always emitted by an RENode, meaning that
 /// there is always an emitter of some [`NodeId`].
 #[derive(Debug, Clone, PartialEq, Eq, Hash, ScryptoSbor)]
-pub struct EventTypeIdentifier(pub Emitter, pub TypePointer);
+pub struct EventTypeIdentifier(pub Emitter, pub String);
```
Comment (iamyulong) [inline radix-engine-interface/src/types/event_id.rs:19]: Nice

File: radix-engine-common/src/types/type_identifier.rs [hunk 2]
```diff
@@ -4,3 +4,8 @@ use sbor::LocalTypeIndex;
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Sbor)]
 pub struct TypeIdentifier(pub Hash, pub LocalTypeIndex);
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Sbor)]
+pub enum TypeSubstitutionRef {
```
Comment (iamyulong) [inline radix-engine-common/src/types/type_identifier.rs:9]: `TypeRefSubstitution`?

File: radix-engine/src/system/payload_validation.rs [hunk 3]
```diff
@@ -26,10 +26,14 @@ use super::system_callback_api::SystemCallbackObject;
...
 /// We use a trait here so it can be implemented either by the System API (mid-execution) or by off-ledger systems
-pub trait TypeInfoLookup {
+pub trait ValidationContext {
     fn get_node_type_info(&self, node_id: &NodeId) -> Result<TypeInfoForValidation, RuntimeError>;
 
     fn schema_origin(&self) -> &SchemaOrigin;
+
+    fn allow_ownership(&self) -> bool;
+
+    fn allow_non_global_ref(&self) -> bool;
```
Comment (iamyulong) [inline radix-engine/src/system/payload_validation.rs:36]: Nice!

**Review by iamyulong — COMMENTED** [review]
Quite like the new payload schema validation!

File: radix-engine-interface/src/blueprints/package/substates.rs [hunk 4]
```diff
@@ -84,16 +100,16 @@ impl Clone for PackageRoyaltyAccumulatorSubstate {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Sbor)]
-pub enum TypePointer {
-    Package(TypeIdentifier), // For static types
-    Instance(u8),            // For generics
+pub enum BlueprintPayloadDef {
+    Static(TypeIdentifier),
+    Generic(u8),
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine-interface/src/blueprints/package/substates.rs:105]: It's not a type but more like the definition of a type (most similar to rust structure `FieldDef`: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.FieldDef.html)

File: radix-engine-common/src/types/type_identifier.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine-common/src/types/type_identifier.rs:9]: It's more like the "reference to the type I'd like to substitute" NOT "substitution of the type reference"

File: radix-engine-interface/src/blueprints/package/substates.rs [hunk 5]
```diff
@@ -238,41 +371,54 @@ impl IndexedStateSchema {
...
+                    BlueprintPayloadDef::Static(TypeIdentifier(schema_hash, type_index))
                 }
-                TypeRef::Generic(instance_index) => TypePointer::Instance(instance_index),
+                TypeRef::Generic(instance_index) => BlueprintPayloadDef::Generic(instance_index),
             });
-            collections.push((PartitionOffset(partition_offset), schema));
-            partition_offset += 1;
+
+            if let Some(partition_num) = system_mappings.get(&collection_index) {
+                collections.push((PartitionDescription::Physical(*partition_num), schema));
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine-interface/src/blueprints/package/substates.rs:403]: Yes, this is probably a good idea

File: radix-engine-interface/src/api/object_api.rs [hunk 6]
```diff
@@ -99,6 +98,12 @@ impl FieldValue {
     }
 }
 
+#[derive(Default, Debug, Clone, PartialEq, Eq, ScryptoSbor)]
+pub struct GenericArgs {
+    pub additional_schema: Option<ScryptoSchema>,
+    pub type_substitution_refs: Vec<GenericSubstitution>,
```
Comment (iamyulong) [inline radix-engine-interface/src/api/object_api.rs:104]: `generic_args` or `generic_substitutions`

File: radix-engine-interface/src/blueprints/package/invocations.rs [hunk 7]
```diff
@@ -297,7 +297,7 @@ impl PackageDefinition {
                             BlueprintKeyValueSchema {
                                 key: TypeRef::Static(LocalTypeIndex::WellKnown(ANY_TYPE)),
                                 value: TypeRef::Static(LocalTypeIndex::WellKnown(ANY_TYPE)),
-                                can_own: true,
+                                allow_ownership: true,
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/package/invocations.rs:300]: We still have some `can_own` in the code base. Worth a global find & replace?

File: radix-engine-interface/src/blueprints/package/substates.rs [hunk 8]
```diff
@@ -203,30 +245,104 @@ impl BlueprintInterface {
...
     }
 
-    pub fn get_function_output_type_pointer(&self, ident: &str) -> Option<TypePointer> {
+    pub fn get_function_output_type_pointer(&self, ident: &str) -> Option<BlueprintPayloadDef> {
         let schema = self.functions.get(ident)?;
         Some(schema.output.clone())
     }
 
-    pub fn get_event_type_pointer(&self, event_name: &str) -> Option<TypePointer> {
+    pub fn get_event_type_pointer(&self, event_name: &str) -> Option<BlueprintPayloadDef> {
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/package/substates.rs:258]: Can we rename all `_type_pointer` to `_payload_def`?

File: radix-engine-interface/src/blueprints/package/substates.rs [hunk 9]
```diff
@@ -298,112 +446,74 @@ impl IndexedStateSchema {
...
+        let (_partition, schema) = self.collections.get(collection_index.clone() as usize)?;
+        match schema {
+            BlueprintCollectionSchema::SortedIndex(index) => Some(index.key.clone()),
             _ => None,
         }
     }
 
-    pub fn into_sorted_index_partition(
-        mut self,
+    pub fn get_sorted_index_type_pointer_value(
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/package/substates.rs:493]: All `_type_pointer` to `_payload_def`

File: radix-engine-queries/src/typed_substate_layout.rs [hunk 10]
```diff
@@ -282,12 +293,14 @@ fn to_typed_object_substate_key_internal(
...
                     )
                 }
+                /*
                 PackagePartitionOffset::Schemas => {
                     let key = substate_key.for_map().ok_or(())?;
                     TypedMainModuleSubstateKey::PackageSchemaKey(
                         scrypto_decode(&key).map_err(|_| ())?,
                     )
                 }
+                 */
```
Comment (iamyulong) [inline radix-engine-queries/src/typed_substate_layout.rs:303]: ?

File: radix-engine/src/blueprints/package/package.rs [hunk 11]
```diff
@@ -680,7 +676,7 @@ pub fn create_bootstrap_package_partitions(
                     blueprint_id: BlueprintId::new(&PACKAGE_PACKAGE, PACKAGE_BLUEPRINT),
                     outer_obj_info: OuterObjectInfo::default(),
                     features: btreeset!(),
-                    instance_schema: None,
+                    type_substitutions_refs: vec![],
```
Comment (iamyulong) [inline radix-engine/src/blueprints/package/package.rs:679]: `generic_substitutions`

File: radix-engine/src/transaction/system_structure.rs [hunk 12]
```diff
@@ -283,34 +293,39 @@ impl<'a, S: SubstateDatabase> SubstateSchemaMapper<'a, S> {
...
+                                    SortedIndexPartitionEntryStructure {
+                                        key_schema: resolver.resolve(kv_schema.key),
+                                        value_schema: resolver.resolve(kv_schema.value),
+                                    },
+                                )
+                            }
+                        }
                     }
                 }
+                PartitionDescription::Physical(..) => {}
```
Comment (iamyulong) [inline radix-engine/src/transaction/system_structure.rs:328]: No resolving for schema?

**Review by iamyulong — APPROVED** [review]

**Pull request merged**